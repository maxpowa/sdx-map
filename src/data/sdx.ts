import { GPSSystem } from '../util/gps'

export const DraconisExpanseSystem = {
  // This GPS list is exported from ingame, using a combination of `!nexus getsectors true` and `!gps` commands.
  // Then run through `parseGPSList` to convert it to a format that can be used here.
  // Some manual tweaks are made as well, marking zone speed limits (by specifying instance names)
  Sol: GPSSystem.fromString(`
    GPS:Ariel:-4110989.31:3110909.45:-359192.17:#FF40EC34:
    GPS:Beratnas Gas Station Jupiter:-119500.8:-2955103.11:351258.18:#FFD6931E:
    GPS:Beratnas Gas Station Saturn:2290226.37:1817047.82:-195774.09:#FFD6931E:
    GPS:Beratnas Gas Station Uranus:-4106322.64:3003556.83:-69235.74:#FFD6931E:
    GPS:Bush Naval Shipyards:-860377.39:251612.55:58803.99:#FF75C9F1:Nation HQs:
    GPS:C1 - (R:300km):600000:1600000:0:#FFFFFF00:DX7:
    GPS:C2 - (R:300km):-600000:1600000:0:#FFFFFF00:DX7:
    GPS:Ceres - (R:400km):0:2350000:100000:#FFFFFF00:DX7:
    GPS:Ceres Station:-9793.86:2325984.53:84772.28:#FFD6931E:
    GPS:Ceres:0:2350000:100000:#FFD6931E:
    GPS:Corley Station:3001481.29:1908187.5:-105165.49:#FFD6931E:
    GPS:Deimos:900000:-800000:0:#FF40EC34:
    GPS:E1 - (R:300km):-1725000:0:0:#FFFFFF00:DX4:
    GPS:E2 - (R:300km):-1500000:-900000:0:#FFFFFF00:DX4:
    GPS:Earth - (R:400km):-800000:100000:80000:#FFFFFF00:DX4:
    GPS:Earth:-800000:100000:80000:#FF40EC34:
    GPS:Europa KOTH:404015.47:-3195511.14:108837.6:#FFD6931E:
    GPS:Europa:400000:-3200000:100000:#FF40EC34:
    GPS:Far Gate Zone - (R:80km):-9999999:-9999999:9999999:#FFFFFF00:DX6:
    GPS:Ganymede Botanical:-404041.3:-3093957.08:-285675.6:#FFD6931E:
    GPS:Ganymede:-400000:-3100000:-300000:#FF40EC34:
    GPS:Greek Cluster - (R:400km):-1600000:-2700000:-334490:#FFFFFF00:DX2:
    GPS:Greek Cluster Waystation:-1380846.75:-2582155.58:-302303.08:#FFD6931E:
    GPS:Inner High Speed Zone - (R:2750km):0:-1725000:100000:#FFFFFF00:DX5:
    GPS:Io:200000:-2900000:-100000:#FF40EC34:
    GPS:Jupiter - (R:750km):0:-3200000:0:#FFFFFF00:DX6:
    GPS:Jupiter:0:-3200000:0:#FF40EC34:
    GPS:Kirino Orbital Shipyards:827693.09:-813745.92:-77878.79:#FFF1C875:Nation HQs:
    GPS:Luna:-650000:225000:0:#FF40EC34:
    GPS:M1 - (R:300km):800000:-1550000:0:#FFFFFF00:DX3:
    GPS:M2 - (R:300km):1550000:-800000:0:#FFFFFF00:DX3:
    GPS:Mars - (R:400km):750000:-750000:-80000:#FFFFFF00:DX3:
    GPS:Mars:750000:-750000:-80000:#FF40EC34:
    GPS:MCRN Free Rebel Fleet:3200000:2250000:35000:#FFD6931E:
    GPS:Miller Mobile Construction Barge:-9605.07:2310850.71:122318.61:#FFF17575:Nation HQs:
    GPS:P1 - (R:300km):-800000:-1550000:0:#FFFFFF00:DX2:
    GPS:Pallas - (R:300km):0:-1725000:100000:#FFFFFF00:DX2:
    GPS:Pallas:0:-1725000:100000:#FFFFFF00:
    GPS:Pallas Station:-3296:-1727232:76613:#FFD6931E:
    GPS:Phobos:650000:-650000:-250000:#FF40EC34:
    GPS:Rhea:2700000:1800000:100000:#FF40EC34:
    GPS:Saturn - (R:750km):2800000:2000000:-100000:#FFFFFF00:DX8:
    GPS:Saturn:2800000:2000000:-100000:#FF40EC34:
    GPS:Scrap Yard:2750000:2435000:130000:#FFD6931E:
    GPS:Sol Gate - (R:80km):-3900000:3600000:-200000:#FFFFFF00:DX6:
    GPS:Sol Gate:-3900000:3600000:-200000:#FFFFFF00:
    GPS:T1 - (R:300km):1725000:0:0:#FFFFFF00:DX2:
    GPS:Titan:3000000:1900000:-100000:#FF40EC34:
    GPS:Trojan Cluster - (R:400km):1600000:-2700000:445350:#FFFFFF00:DX2:
    GPS:Trojan Cluster Waystation:1420788.01:-2531349.34:404487.57:#FFD6931E:
    GPS:Tycho - (R:300km):1400000:1000000:-50000:#FFFFFF00:DX2:
    GPS:Tycho Hub:1400000:1000000:-50000:#FFD6931E:
    GPS:Uranus - (R:500km):-4200000:3000000:-250000:#FFFFFF00:DX6:
    GPS:Uranus:-4200000:3000000:-250000:#FF40EC34:
    GPS:Vesta - (R:300km):-1400000:1000000:-50000:#FFFFFF00:DX2:
    GPS:Vesta Station:-1412860:986040:-49577:#FFD6931E:
    GPS:Vesta:-1400000:1000000:-50000:#FFD6931E:
  `),
  Kronos: GPSSystem.fromString(`
    GPS:Kronos Gate:-35303:234184:198247:#FFD6931E:
    GPS:Kronos:-75000:86000:18000:#FF40EC34:
    GPS:Kronos 1:-224000:164000:-55000:#FF40EC34:
    GPS:Kronos Research Facility:-48000:29000:32000:#FFD6931E:
  `),
  Ilus: GPSSystem.fromString(`
    GPS:Ilus:69500:111150:58750:#FFB775F1:
    GPS:Ilus Gate:300146:375140:31503:#FFB775F1:
    GPS:Ilus 1:57500:318500:47900:#FF40EC34:
    GPS:Belter Colony:71459:149570:12666:#FFB775F1:
  `),
  Jannah: GPSSystem.fromString(`
    GPS:Alpha Berkut Station:600764.75:703346.87:500064.02:#FF40EC34:
    GPS:Port Jannah:422538.91:630064.11:135873.58:#FF40EC34:
    GPS:Jannah Gate:250000:340000:25000:#FFD6931E:
    GPS:JN1 Asteroid Cluster:600000:700000:600000:#FF40EC34:
    GPS:Jannah:600000:700000:100000:#FF40EC34:
  `),
  'Ring Space': GPSSystem.fromString(`
    GPS:Medina Station:15156.6:11584.25:2312.33:#FFF1A875:
    GPS:Sol Gate:-17:411:18:#FFF1A875:
    GPS:Ilus Gate:12349:28361:7116:#FFF1A875:
    GPS:Kronos Gate:32676:6550:1022:#FFF1A875:
  `),
}

// Transit points are used to define the possible destinations from a given system, for calculating inter-system travel times.
export const TransitPoints = {
  'Jannah Gate': ['Sol'],
  'Sol Gate': ['Ring Space', 'Sol'],
  'Ilus Gate': ['Ring Space', 'Ilus'],
  'Kronos Gate': ['Ring Space', 'Kronos'],
}
