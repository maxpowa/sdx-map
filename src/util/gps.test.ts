import { describe, it, expect } from 'vitest'
import {
  GPSSystem,
  GPSPoint,
  GPSBody,
  GPSZone,
  computeShortestRoute,
} from './gps'
import { Vector3 } from 'three'
import { DraconisExpanseSystem } from '../data/sdx'

describe('GPS module', () => {
  it('should parse basic GPS data', () => {
    const gps = 'GPS:Hello World:15:-15:0:#40EC34:'
    const result = GPSPoint.fromString(gps)
    expect(result).toEqual({
      name: 'Hello World',
      x: 15,
      y: -15,
      z: 0,
      color: '#40EC34',
      class: 'poi',
    })
  })

  it('should parse GPS data with radius', () => {
    const gps = 'GPS:Hello World - (R:10km):15:-15:0:#40EC34:'
    const result = GPSZone.fromString(gps)
    expect(result).toEqual({
      name: 'Hello World - (R:10km)',
      x: 15,
      y: -15,
      z: 0,
      color: '#40EC34',
      class: 'zone',
      radius: 10000,
      childZones: [],
      children: [],
    })
  })

  it('should parse GPS data with radius and category-based color', () => {
    const gps = 'GPS:Hello World - (R:10km):15:-15:0:#40EC34:DX2:'
    const result = GPSZone.fromString(gps)
    expect(result).toEqual({
      name: 'Hello World - (R:10km)',
      x: 15,
      y: -15,
      z: 0,
      category: 'DX2',
      color: '#FF8C00',
      class: 'zone',
      radius: 10000,
      childZones: [],
      children: [],
    })
  })

  it('should parse GPS data with known body', () => {
    const gps = 'GPS:Earth:15:-15:0:#40EC34:'
    const result = GPSBody.fromString(gps)
    expect(result).toEqual({
      name: 'Earth',
      x: 15,
      y: -15,
      z: 0,
      category: undefined,
      color: '#40EC34',
      class: 'body',
      radius: 100000,
    })
  })

  it('should be reasonably robust against colons in GPS names', () => {
    const gps = 'GPS:Hello:World:15:-15:0:#40EC34:'
    const result = GPSPoint.fromString(gps)
    expect(result).toEqual({
      category: undefined,
      class: 'poi',
      color: '#40EC34',
      name: 'Hello:World',
      x: 15,
      y: -15,
      z: 0,
    })
  })

  it('should throw an error for invalid GPS data', () => {
    const gps = 'GPS:Hell:o:World:1::5:-:1:::5:0:#FF40:EC34'
    expect(() => GPSPoint.fromString(gps)).toThrowError(
      'Failed to parse GPS, invalid format.',
    )
  })

  it('should export basic GPS data to string', () => {
    const gps = 'GPS:Hello World:15:-15:0:#FF40EC34:'
    const result = GPSPoint.fromString(gps).toString()
    expect(result).toEqual(gps)
  })

  it('should parse GPS list data', () => {
    const gps = `GPS:Hello World:15:-15:0:#40EC34:\nGPS:Goodbye World:-15:15:0:#40EC34:`
    const result = GPSSystem.fromString(gps)
    expect(result).toMatchObject({
      class: 'system',
      color: '#FFFFFF',
      name: 'System - (R:-0.001km)',
      parent: undefined,
      radius: -1,
      x: 0,
      y: 0,
      z: 0,
      childZones: [],
      children: [
        {
          name: 'Hello World',
          x: 15,
          y: -15,
          z: 0,
          color: '#40EC34',
          class: 'poi',
        },
        {
          name: 'Goodbye World',
          x: -15,
          y: 15,
          z: 0,
          color: '#40EC34',
          class: 'poi',
        },
      ],
    })
  })

  it('should export GPS list data to string', () => {
    const gps = `GPS:Hello World:15:-15:0:#FF40EC34:\nGPS:Goodbye World:-15:15:0:#FF40EC34:`
    const result = GPSSystem.fromString(gps).toString()
    expect(result).toEqual(gps)
  })

  it('should calculate distances between GPS points', () => {
    const pointA = new GPSPoint(0, 0, 0, 'poi', 'A', '#FFFFFF')
    const pointB = new GPSPoint(3, 4, 0, 'poi', 'B', '#FFFFFF')
    const result = pointA.distanceTo(pointB)
    expect(result).toEqual(5)
  })

  it('should calculate distances between GPS points with different z values', () => {
    const pointA = new GPSPoint(0, 0, 0, 'poi', 'A', '#FFFFFF')
    const pointB = new GPSPoint(3, 4, 5, 'poi', 'B', '#FFFFFF')
    const result = pointA.distanceTo(pointB)
    expect(result).toEqual(7.0710678118654755)
  })

  it('should calculate the shortest vector to the edge of a zone', () => {
    const zone = new GPSZone(2, 10, 7, 'A', '#FFFFFF', 10000)
    const point = new GPSPoint(0, 500, 0, 'poi', 'B', '#FFFFFF')
    expect(point.distanceTo(zone)).toEqual(490.05407864846916)
    const result = zone.vectorToEdge(point)
    expect(result).toEqual(
      new Vector3(-38.811822350623906, 10008.896475902857, -135.8413782271837),
    )
  })

  it('should calculate turns for a GPS point within a zone', () => {
    const system = GPSSystem.fromString(`
      GPS:Deimos:900000:-800000:0:#40EC34:
      GPS:Phobos:650000:-650000:-250000:#40EC34:
      GPS:Mars:750000:-750000:-80000:#40EC34:
      GPS:Mars - (R:400km):750000:-750000:-80000:#FFFFFF00:DX3:
    `)
    expect(system.turns().length).toEqual(2)
  })
})

describe('Route Planner', () => {
  it('should calculate the optimal route between two points with an obstacle', () => {
    const world = GPSSystem.fromString(`
      GPS:Deimos:900000:-800000:0:#40EC34:
      GPS:Phobos:650000:-650000:-250000:#40EC34:
      GPS:Mars:750000:-750000:-80000:#40EC34:
      GPS:Mars - (R:400km):750000:-750000:-80000:#FFFFFF00:DX3:
    `)
    const deimos = world.pois(true)[0]
    const phobos = world.pois(true)[1]
    const result = computeShortestRoute([deimos, phobos], world)
    expect(result).toMatchSnapshot()
  })

  it('should calculate a complex route', () => {
    const globalData = DraconisExpanseSystem['Sol']
    const deimos = globalData
      .pois(true)
      .find((body) => body.name === 'Deimos') as GPSPoint
    const ariel = globalData
      .pois(true)
      .find((body) => body.name === 'Ariel') as GPSPoint

    const route = computeShortestRoute([deimos, ariel], globalData)
    expect(route).toMatchSnapshot()
  })
})
