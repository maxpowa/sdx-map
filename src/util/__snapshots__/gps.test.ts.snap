// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Route Planner > should calculate a complex route 1`] = `
[
  GPSPoint {
    "category": "slowzone",
    "class": "poi",
    "color": "#40EC34",
    "name": "Deimos",
    "parent": undefined,
    "radius": undefined,
    "x": 900000,
    "y": -800000,
    "z": 0,
  },
  GPSPoint {
    "category": "highspeed",
    "class": "poi",
    "color": "#40EC34",
    "name": "Turn (Deimos)",
    "parent": undefined,
    "radius": undefined,
    "x": 1089869.3372478222,
    "y": -863289.7790826075,
    "z": 101263.64653217181,
  },
  GPSPoint {
    "category": "highspeed",
    "class": "poi",
    "color": "#FF0000",
    "name": "Obstacle (Mars - (R:400km))",
    "parent": undefined,
    "radius": undefined,
    "x": 1074471.3741052279,
    "y": -725346.2784975092,
    "z": 230174.86047575067,
  },
  GPSPoint {
    "category": "highspeed",
    "class": "poi",
    "color": "#FF0000",
    "name": "Obstacle (Mars - (R:400km))",
    "parent": undefined,
    "radius": undefined,
    "x": 893114.0694867821,
    "y": -522459.5884340666,
    "z": 280510.32457491197,
  },
  GPSPoint {
    "category": "highspeed",
    "class": "poi",
    "color": "#FF8C00",
    "name": "Obstacle (Vesta - (R:300km))",
    "parent": undefined,
    "radius": undefined,
    "x": -1218963.1149383665,
    "y": 1270334.8283263796,
    "z": 78797.65686730255,
  },
  GPSPoint {
    "category": "highspeed",
    "class": "poi",
    "color": "#008000",
    "name": "Obstacle (Uranus - (R:500km))",
    "parent": undefined,
    "radius": undefined,
    "x": -3889118.8454798968,
    "y": 3313895.103712005,
    "z": -577576.5005835672,
  },
  GPSPoint {
    "category": "slowzone",
    "class": "poi",
    "color": "#FF8C00",
    "name": "Lithoturn (Uranus - (R:500km))",
    "parent": undefined,
    "radius": undefined,
    "x": -3952090.922759164,
    "y": 3307845.5164868473,
    "z": -553345.2344072373,
  },
  GPSPoint {
    "category": "slowzone",
    "class": "poi",
    "color": "#40EC34",
    "name": "Ariel",
    "parent": undefined,
    "radius": undefined,
    "x": -4105528.3827548507,
    "y": 3117713.896041813,
    "z": -365891.25857138366,
  },
]
`;

exports[`Route Planner > should calculate the optimal route between two points with an obstacle 1`] = `
[
  GPSPoint {
    "category": "slowzone",
    "class": "poi",
    "color": "#40EC34",
    "name": "Deimos",
    "parent": undefined,
    "radius": undefined,
    "x": 900000,
    "y": -800000,
    "z": 0,
  },
  GPSPoint {
    "category": "slowzone",
    "class": "poi",
    "color": "#40EC34",
    "name": "Obstacle (Mars)",
    "parent": undefined,
    "radius": undefined,
    "x": 818756.194359412,
    "y": -722843.175786474,
    "z": -130962.25209164093,
  },
  GPSPoint {
    "category": "slowzone",
    "class": "poi",
    "color": "#40EC34",
    "name": "Phobos",
    "parent": undefined,
    "radius": undefined,
    "x": 655622.6273413141,
    "y": -652418.8335137389,
    "z": -246048.17271802254,
  },
]
`;
